prefetch.bs

```diff
@@ -0,0 +1,219 @@
+    1. Let |locationURL| be null.
+    1. Let |currentURL| be |request|'s [=request/current URL=].
+    1. While true:
+        1. If |locationURL| is non-null, then set |currentURL| to |locationURL|.
+        1. If |request|'s [=request/reserved client=] is not null and |currentURL|'s [=url/origin=] is not the [=same origin|same=] as |request|'s [=request/reserved client=]'s [=environment/creation URL=]'s [=url/origin=], then:
+            1. Run the [=environment discarding steps=] for |request|'s [=request/reserved client=].
+            1. Set |request|'s [=request/reserved client=] to null.
+        1. If |request|'s [=request/reserved client=] is null, then:
+            1. Let |topLevelCreationURL| be |currentURL|.
+            1. Let |topLevelOrigin| be null.
+            1. If |browsingContext| is not a [=top-level browsing context=], then:
+                1. Let |parentEnvironment| be |browsingContext|'s [=browsing context/container=]'s [=relevant settings object=].
+                1. Set |topLevelCreationURL| to |parentEnvironment|'s [=environment/top-level creation URL=] and |topLevelOrigin| to |parentEnvironment|'s [=environment/top-level origin=].
+            1. Set |request|'s [=request/reserved client=] to a new [=environment=] whose [=environment/id=] is a unique opaque string, [=environment/target browsing context=] is |browsingContext|, [=environment/creation URL=] is |currentURL|, [=environment/top-level creation URL=] is |topLevelCreationURL|, and [=environment/top-level origin=] is |topLevelOrigin|.
+        1. Let |proposedPartitionKey| be the result of [=determining the network partition key=] given |request|.
+        1. If |partitionKey| is not equal to |proposedPartitionKey|, then set |response| to a [=network error=] and [=iteration/break=].
+
+            <div class="issue">It might be possible to "downgrade" to [=uncredentialed prefetch=] in this case.</div>
+        1. TODO: this is where [=process a navigate fetch=] checks the `navigate-to` CSP
+        1. If |response| is null, [=fetch=] |request|.
+        1. Otherwise, perform [=HTTP-redirect fetch=] using |request| and |response|.
+        1. Wait for the [=task=] on the [=networking task source=] to [=process response=] and set |response| to the result.
+        1. TODO: Figure out if the coop enforcement logic is necessary here; it probably is
+        1. TODO: Figure out if the corp enforcement logic is necessary here; it probably is
+        1. Set |locationURL| to |response|'s [=response/location URL=].
+        1. If |locationURL| is not a [=URL=] whose [=url/scheme=] is an [=HTTP(S) scheme=], then [=iteration/break=].
+    1. If |locationURL| is failure or a [=URL=] whose [=url/scheme=] is not an [=HTTP(S) scheme=], then set |response| to a [=network error=].
+    1. TODO: navigate-to, frame-src, XFO enforcement should probably be left to navigation, but what about status codes and Content-Disposition?
```

Comment:
Would it be observable either way? I guess maybe if a server was going to give a 204 the first time (during prefetch) and a 200 the second time (during navigation), then caching the 204 in the prefetch buffer would be a bit suboptimal?

I mean, do we even want to cache 404s in the prefetch buffer?

Reply:
Would have to poke at what classic prefetch does. At the moment the /prefetch_proxy/ implementation requires a 2XX status code and text/html mime type. It requires that SimpleURLLoader successfully produce a body, but it's not obvious to me without trying it whether a 204/205 results in an empty or absent body string in that code path.

My gut is that we probably do want to require an OK status and possibly just 200. OTOH, I don't see as much reason to require `text/html`, but Ryan would know whether there was a concrete reason behind that.

Non-HTML responses make `Content-Disposition` a little more interesting even though it's not really the marquee use case. I suspect actually handing off to the Chromium download manager would be a mild implementation headache, but I don't really see a conceptual reason it shouldn't be legal.

Reply:
I think OK status that is not 204 or 205 is probably the right call. Content type... hmm. I mean, I like enforcing good content type hygiene in general, but we should at least support `application/xhtml+xml` or whatever the XHTML content type is.

I think avoiding downloads or non-HTML pages is reasonable just as a simplification effort, since the benefits of prerendering them are pretty marginal.

Reply:
Especially in the prefetch case, it doesn't seem unreasonable that a user might want a tax form PDF or something like that to load quickly, for instance.